### PR TITLE
[ML] allow model_aliases to be used with Pytorch trained models

### LINKED ISSUE #90960: [ML] Allow model aliases to be used for native (PyTorch) models
### Description

Currently, model aliases are disallowed for native models. This was mainly due to some more complicated logic within the ModelLoader that attempted to auto-load models based on alias changes. 

But, this should not mean that we always disallow model aliases for PyTorch models. We just need to be careful about the validation of the changes and make sure the modelLoader doesn't break.

 - Make sure that model alias changes for native models don't break the ModelLoader
 - Enforce model alias change rules for changes affecting native models:
    - Disallow model alias deletions when model is deployed
    - Disallow changing the model alias from a deployed model to an undeployed one
    - Disallow changing the model alias from one NLP task type to another

### PR DESCRIPTION
**Pull request by benwtrent**
This adds model_alias support for native pytorch models.

Model aliases can be used in `_infer` or within the inference processor. This way the alias can be atomically changed without down time to another deployed model. 

Restrictions:
 - Model alias changes need to be done between two models of the same kind (e.g. pytorch -> pytorch)
 - Model alias change is not allowed between a model that is deployed to a model that is not
 - Model alias change is not allowed between a model that deployed AND allocated to a model that is deployed but NOT allocated (not assigned to any nodes).
 - A deployment cannot be stopped (without supplying the `force` parameter) when the model has a model alias that is used by a pipeline.

closes: https://github.com/elastic/elasticsearch/issues/90960

### CODE REVIEW

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Review by davidkyle — APPROVED** [review]
LGTM

Please can you update `put-trained-models-aliases.asciidoc` with the restrictions you documented in the PR description

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/3rd_party_deployment.yml [hunk 1]
```diff
@@ -278,3 +316,66 @@ setup:
...
+        model_id: "pytorch"
+        body: >
+          {
+            "docs": [
+              { "input": "words" }
+            ]
+          }
+  - do:
+      ml.get_trained_models_stats:
+        model_id: "test_model"
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/3rd_party_deployment.yml:372]: @davidkyle it does.

[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/3rd_party_deployment.yml:372]: @davidkyle it has always been that way, so, removing the pytorch restriction gives us this.

**Comment by benwtrent** [comment]
@elastic/ml-docs ping for review :)

File: docs/reference/ml/trained-models/apis/infer-trained-model-deployment.asciidoc [hunk 2]
```diff
@@ -31,7 +31,7 @@ deprecated::[8.3.0,Replaced by <<infer-trained-model>>.]
 
 `<model_id>`::
 (Required, string)
-include::{es-repo-dir}/ml/ml-shared.asciidoc[tag=model-id]
+include::{es-repo-dir}/ml/ml-shared.asciidoc[tag=model-id-or-alias]
```
Comment (szabosteve) [inline docs/reference/ml/trained-models/apis/infer-trained-model-deployment.asciidoc:34]:
```suggestion
include::{es-repo-dir}/ml/ml-shared.asciidoc[tag=model-id]
```

File: docs/reference/ml/trained-models/apis/put-trained-models-aliases.asciidoc [hunk 3]
```diff
@@ -34,11 +34,17 @@ and processors.
...
++
+--
+* You are not allowed to update an alias such that it references a different
+trained model ID and the model uses a different type of {dfanalytics}. For example,
+this situation occurs if you have a trained model for
 {reganalysis} and a trained model for {classanalysis}; you cannot reassign an
 alias from one type of trained model to another.
+* You cannot update an alias from a `pytorch` model and a {dfanalytics} model.
+* You cannot update the alias from a deployed `pytorch` model to one
+not currently deployed.
```
Comment (szabosteve) [inline docs/reference/ml/trained-models/apis/put-trained-models-aliases.asciidoc:47]: It'll fix the docs build.

```suggestion
not currently deployed.
--
```

**Review by szabosteve — APPROVED** [review]
One suggestion that fixes the build. Otherwise LGTM, thanks!

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Pull request merged**